File: tests/bench/fib_recursive.cpp
// Same program as fib_recursive.jai.
#include <cstdint>
#include <cstdio>

#include <cstdlib>
#include <cstring>

// BENCH_LEVEL picks how much work this benchmark does: easy a sixteenth of it,
// medium a quarter, hard (the default, and anything unrecognised) all of it.
static long bench_scale() {
    const char *l = std::getenv("BENCH_LEVEL");
    if (l && std::strcmp(l, "easy") == 0) return 16;
    if (l && std::strcmp(l, "medium") == 0) return 4;
    return 1;
}
static const long SCALE = bench_scale();
// Work is exponential in n, so shorten it by subtracting.
static const int64_t DEPTH_DROP = SCALE == 16 ? 4 : (SCALE == 4 ? 2 : 0);

static int64_t fib(int64_t n) { return n < 2 ? n : fib(n - 1) + fib(n - 2); }

int main() {
    std::printf("%lld\n", (long long)fib(38 - DEPTH_DROP));
    return 0;
}

File: tests/bench/fib_recursive.py
import os

_LEVEL = os.environ.get("BENCH_LEVEL", "hard")
# Work is exponential in n, so the level subtracts from it instead of dividing.
DEPTH_DROP = 4 if _LEVEL == "easy" else 2 if _LEVEL == "medium" else 0
N = 38 - DEPTH_DROP


def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

print(fib(N))
